libs/process/stub.go

```diff
@@ -0,0 +1,149 @@
+package process
+
+import (
+	"context"
+	"fmt"
+	"os/exec"
+	"path/filepath"
+	"strings"
+)
+
+var stubKey int
+
+// WithStub creates process stub for fast and flexible testing of subprocesses
+func WithStub(ctx context.Context) (context.Context, *processStub) {
+	stub := &processStub{responses: map[string]reponseStub{}}
+	ctx = context.WithValue(ctx, &stubKey, stub)
+	return ctx, stub
+}
+
+func runCmd(ctx context.Context, cmd *exec.Cmd) error {
+	stub, ok := ctx.Value(&stubKey).(*processStub)
+	if ok {
+		return stub.run(cmd)
+	}
+	return cmd.Run()
+}
+
+type reponseStub struct {
+	stdout string
+	stderr string
+	err    error
+}
+
+type processStub struct {
+	reponseStub
+	calls           []*exec.Cmd
+	defaultCallback func(*exec.Cmd) error
+	responses       map[string]reponseStub
+}
+
+func (s *processStub) WithDefaultOutput(output string) *processStub {
+	s.reponseStub.stdout = output
+	return s
+}
+
+func (s *processStub) WithDefaultFailure(err error) *processStub {
+	s.reponseStub.err = err
+	return s
+}
+
+func (s *processStub) WithDefaultCallback(cb func(cmd *exec.Cmd) error) *processStub {
+	s.defaultCallback = cb
+	return s
+}
+
+// WithStdoutFor predefines standard output response for a command. The first word
+// in the command string is the executable name, and NOT the executable path.
+// The following command would stub "2" output for "/usr/local/bin/echo 1" command:
+//
+//	stub.WithStdoutFor("echo 1", "2")
+func (s *processStub) WithStdoutFor(command, out string) *processStub {
+	s.responses[command] = reponseStub{
+		stdout: out,
+	}
```

Comment:
This overwrites existing entries so that composition is not possible. Is that conscious or an oversight?

Reply:
conscious